File: app/src/main/java/com/handwave/swipe/MainActivity.java
package com.handwave.swipe;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

public class MainActivity extends Activity {
    private static final int CAMERA_REQ = 100;
    private static final int NOTIFICATION_REQ = 101;
    private TextView status;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        buildUi();
        requestNotificationPermissionIfNeeded();
    }

    private void buildUi() {
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(48, 56, 48, 40);

        TextView title = new TextView(this);
        title.setText("HandSwipe");
        title.setTextSize(30);
        title.setTextColor(Color.BLACK);

        TextView info = new TextView(this);
        info.setText("Camera trước nhận diện chuyển động tay.\nTay đi lên = vuốt lên. Tay đi xuống = vuốt xuống.");
        info.setTextSize(16);
        info.setPadding(0, 24, 0, 24);

        status = new TextView(this);
        status.setTextSize(16);
        status.setPadding(0, 0, 0, 24);

        Button accessibility = new Button(this);
        accessibility.setText("1. Bật quyền Trợ năng");
        accessibility.setOnClickListener(v -> startActivity(new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS)));

        Button start = new Button(this);
        start.setText("2. Bắt đầu");
        start.setOnClickListener(v -> startHandSwipe());

        Button stop = new Button(this);
        stop.setText("Dừng");
        stop.setOnClickListener(v -> stopService(new Intent(this, HandCameraService.class)));

        root.addView(title);
        root.addView(info);
        root.addView(status);
        root.addView(accessibility);
        root.addView(start);
        root.addView(stop);
        setContentView(root);
        updateStatus();
    }

    private void requestNotificationPermissionIfNeeded() {
        if (Build.VERSION.SDK_INT >= 33 && checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.POST_NOTIFICATIONS}, NOTIFICATION_REQ);
        }
    }

    private void startHandSwipe() {
        if (checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.CAMERA}, CAMERA_REQ);
            return;
        }
        if (!SwipeAccessibilityService.isEnabled()) {
            status.setText("Chưa bật Trợ năng. Hãy bật HandSwipe rồi quay lại.");
            startActivity(new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS));
            return;
        }
        Intent service = new Intent(this, HandCameraService.class);
        if (Build.VERSION.SDK_INT >= 26) startForegroundService(service); else startService(service);
        status.setText("ĐANG CHẠY — đưa một bàn tay trước camera.");
    }

    private void updateStatus() {
        status.setText(SwipeAccessibilityService.isEnabled()
                ? "Trợ năng: ĐÃ BẬT"
                : "Trợ năng: CHƯA BẬT");
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (status != null) updateStatus();
    }
}

File: app/src/main/java/com/example/tiktokgesture/MainActivity.kt
package com.example.tiktokgesture

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.provider.Settings
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.example.tiktokgesture.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private var serviceRunning = false

    private val requestCameraPermission =
        registerForActivityResult(androidx.activity.result.contract.ActivityResultContracts.RequestPermission()) { granted ->
            if (granted) {
                Toast.makeText(this, "Đã cấp quyền camera", Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this, "Cần quyền camera để app hoạt động", Toast.LENGTH_LONG).show()
            }
            updateStatus()
        }

    private val requestNotificationPermission =
        registerForActivityResult(androidx.activity.result.contract.ActivityResultContracts.RequestPermission()) { granted ->
            if (!granted) {
                Toast.makeText(
                    this,
                    "Chưa cấp quyền thông báo — service vẫn chạy được nhưng bạn sẽ không thấy thông báo nền",
                    Toast.LENGTH_LONG
                ).show()
            }
        }

    /** Trả về true nếu đã sẵn sàng (đã có quyền hoặc không cần xin trên phiên bản Android này) */
    private fun ensureNotificationPermission() {
        if (android.os.Build.VERSION.SDK_INT >= 33) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS)
                != PackageManager.PERMISSION_GRANTED
            ) {
                requestNotificationPermission.launch(Manifest.permission.POST_NOTIFICATIONS)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.btnGrantCamera.setOnClickListener {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                != PackageManager.PERMISSION_GRANTED
            ) {
                requestCameraPermission.launch(Manifest.permission.CAMERA)
            } else {
                Toast.makeText(this, "Camera đã được cấp quyền rồi", Toast.LENGTH_SHORT).show()
            }
        }

        binding.btnOpenAccessibility.setOnClickListener {
            startActivity(Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS))
            Toast.makeText(
                this,
                "Tìm 'TikTok Gesture' trong danh sách và bật lên",
                Toast.LENGTH_LONG
            ).show()
        }

        binding.btnToggleService.setOnClickListener {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                != PackageManager.PERMISSION_GRANTED
            ) {
                Toast.makeText(this, "Vui lòng cấp quyền camera trước (bước 1)", Toast.LENGTH_LONG).show()
                return@setOnClickListener
            }

            ensureNotificationPermission()

            val intent = Intent(this, GestureForegroundService::class.java)
            if (!serviceRunning) {
                try {
                    ContextCompat.startForegroundService(this, intent)
                    serviceRunning = true
                    Toast.makeText(this, "Đã bật nhận diện cử chỉ. Mở TikTok và thử vẫy tay!", Toast.LENGTH_LONG).show()
                } catch (e: Exception) {
                    Toast.makeText(this, "Lỗi khi bật service: ${e.message}", Toast.LENGTH_LONG).show()
                }
            } else {
                stopService(intent)
                serviceRunning = false
                Toast.makeText(this, "Đã tắt nhận diện cử chỉ", Toast.LENGTH_SHORT).show()
            }
            updateStatus()
        }

        updateStatus()
        ensureNotificationPermission()
    }

    override fun onResume() {
        super.onResume()
        updateStatus()
    }

    private fun updateStatus() {
        val cameraGranted = ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) ==
            PackageManager.PERMISSION_GRANTED
        val accessibilityOn = GestureAccessibilityService.instance != null

        binding.tvStatus.text = buildString {
            append("Camera: ${if (cameraGranted) "✓ đã cấp" else "✗ chưa cấp"}\n")
            append("Accessibility: ${if (accessibilityOn) "✓ đã bật" else "✗ chưa bật"}\n")
            append("Nhận diện cử chỉ: ${if (serviceRunning) "✓ đang chạy" else "✗ đang tắt"}")
        }
    }
}
